server/bleep/src/semantic/embedder.rs

```diff
@@ -65,89 +59,261 @@ pub struct EmbedChunk {
+                permits: Arc::new(tokio::sync::Semaphore::new(session_count)),
+            })
+        }
     }
 
-    async fn batch_embed(&self, log: Vec<&str>) -> anyhow::Result<Vec<Embedding>> {
-        log.into_iter()
-            .map(|entry| tokio::task::block_in_place(|| self.embed(entry)))
-            .collect::<anyhow::Result<Vec<Embedding>>>()
+    #[async_trait]
+    impl Embedder for LocalEmbedder {
+        async fn embed(&self, sequence: &str) -> anyhow::Result<Embedding> {
+            let mut output_request = llm::OutputRequest {
+                all_logits: None,
+                embeddings: Some(Vec::new()),
+            };
+            let vocab = self.model.tokenizer();
+            let beginning_of_sentence = true;
+            let query_token_ids = vocab
+                .tokenize(sequence, beginning_of_sentence)
+                .unwrap()
+                .iter()
+                .map(|(_, tok)| *tok)
+                .collect::<Vec<_>>();
+
+            if let Ok(_permit) = self.permits.acquire().await {
+                for s in &self.sessions {
+                    if let Ok(mut session) = s.try_lock() {
+                        self.model
+                            .evaluate(&mut session, &query_token_ids, &mut output_request);
+                        return Ok(output_request.embeddings.unwrap());
+                    }
+                }
+            }
+
+            unreachable!();
+        }
+
+        fn tokenizer(&self) -> &Tokenizer {
+            &self.tokenizer
+        }
+
+        async fn batch_embed(&self, log: Vec<&str>) -> anyhow::Result<Vec<Embedding>> {
+            let mut output_request = llm::OutputRequest {
+                all_logits: None,
+                embeddings: Some(Vec::new()),
+            };
+            let vocab = self.model.tokenizer();
+            let beginning_of_sentence = true;
+            let query_token_ids = log
+                .iter()
+                .map(|sequence| {
+                    vocab
+                        .tokenize(&sequence, beginning_of_sentence)
+                        .unwrap()
+                        .iter()
+                        .map(|(_, tok)| *tok)
+                        .collect::<Vec<_>>()
+                })
+                .collect::<Vec<_>>();
+            let query_token_ids: Vec<_> = query_token_ids.iter().map(AsRef::as_ref).collect();
+
+            if let Ok(_permit) = self.permits.acquire().await {
+                for s in &self.sessions {
+                    if let Ok(mut session) = s.try_lock() {
+                        self.model.batch_evaluate(
+                            &mut session,
+                            &query_token_ids,
+                            &mut output_request,
+                        );
+                        let embedding: Vec<Vec<f32>> = output_request
+                            .embeddings
+                            .unwrap()
+                            .chunks(384)
```

Comment:
This should be set to `EMBEDDING_DIM`: https://github.com/BloopAI/bloop/blob/7e2cf6bcb207931261ddc962090dda852ab997c4/server/bleep/src/semantic/schema.rs#L12